homeassistant/components/light/lifx.py

```diff
@@ -199,57 +176,111 @@ def supported_features(self):
+
+    @callback
+    def unblock_updates(self, now):
+        """Allow async_update after the new state has settled on the bulb."""
+        self.blocker = None
+        self.schedule_update_ha_state(force_refresh=True)
+
+    def update_later(self, when):
+        """Block immediate update requests and schedule one for later."""
+        if self.blocker:
+            self.blocker()
+        self.blocker = async_track_point_in_utc_time(
+            self.hass, self.unblock_updates,
+            util.dt.utcnow() + timedelta(milliseconds=BULB_LATENCY))
+
+        if self.postponed_update:
+            self.postponed_update()
+        if when > BULB_LATENCY:
+            self.postponed_update = async_track_point_in_utc_time(
+                self.hass, self.update_after_transition,
+                util.dt.utcnow() + timedelta(milliseconds=when+BULB_LATENCY))
+
+    @asyncio.coroutine
+    def async_turn_on(self, **kwargs):
         """Turn the device on."""
         if ATTR_TRANSITION in kwargs:
             fade = int(kwargs[ATTR_TRANSITION] * 1000)
         else:
             fade = 0
 
+        changed_color = False
+
         if ATTR_RGB_COLOR in kwargs:
             hue, saturation, brightness = \
                 convert_rgb_to_hsv(kwargs[ATTR_RGB_COLOR])
+            changed_color = True
         else:
             hue = self._hue
             saturation = self._sat
             brightness = self._bri
 
         if ATTR_BRIGHTNESS in kwargs:
             brightness = kwargs[ATTR_BRIGHTNESS] * (BYTE_MAX + 1)
+            changed_color = True
         else:
             brightness = self._bri
 
         if ATTR_COLOR_TEMP in kwargs:
             kelvin = int(color_temperature_mired_to_kelvin(
                 kwargs[ATTR_COLOR_TEMP]))
+            changed_color = True
         else:
             kelvin = self._kel
 
+        hsbk = [hue, saturation, brightness, kelvin]
         _LOGGER.debug("turn_on: %s (%d) %d %d %d %d %d",
-                      self._ip, self._power,
-                      hue, saturation, brightness, kelvin, fade)
+                      self.ipaddr, self._power, fade, *hsbk)
 
         if self._power == 0:
-            self._liffylights.set_color(self._ip, hue, saturation,
-                                        brightness, kelvin, 0)
-            self._liffylights.set_power(self._ip, 65535, fade)
+            if changed_color:
+                self.device.set_color(hsbk, None, 0)
+            self.device.set_power(True, None, fade)
         else:
-            self._liffylights.set_color(self._ip, hue, saturation,
-                                        brightness, kelvin, fade)
+            self.device.set_power(True, None, 0)     # racing for power status
+            if changed_color:
+                self.device.set_color(hsbk, None, fade)
 
-    def turn_off(self, **kwargs):
+        self.update_later(0)
+        if fade < BULB_LATENCY:
+            self.set_power(1)
```

Comment:
For other reviewers:

This is not doing I/O in the event loop. Checked it, aiolifx will schedule a task on the event loop when we call this. https://github.com/frawau/aiolifx/blob/master/aiolifx/aiolifx34.py#L143